File: 20241022/homin/퍼즐게임.js
function solution(diffs, times, limit) {
  let minLevel = Infinity;
  let maxLevel = -Infinity;

  diffs.forEach((e) => {
    minLevel = Math.min(minLevel, e);
    maxLevel = Math.max(maxLevel, e);
  });

  const argumentSearch = (level) => {
    let sum = 0;
    for (let i = 0; i < diffs.length; i++) {
      if (level >= diffs[i]) {
        sum += times[i];
      } else {
        sum += (diffs[i] - level) * (times[i - 1] + times[i]) + times[i];
      }

      if (sum > limit) {
        return false;
      }
    }

    return limit >= sum ? true : false;
  };

  while (minLevel <= maxLevel) {
    const middle = Math.floor((minLevel + maxLevel) / 2);
    if (argumentSearch(middle)) {
      maxLevel = middle - 1;
    } else {
      minLevel = middle + 1;
    }
  }

  return minLevel;
}

// 1. diff퍼즐의 난이도
// 2. 현재 퍼즐 소요 시간 cur
// 3. 이전 퍼즐 소요 시간 prev
// 4. 숙련도 level
// ----

// 1. 숙련도 >= 난이도  => 0회 틀림 // cur시간 소요
// 2. 반대 => (diff - level)회 틀림 // 틀리면 cur + prev 시간을 추가로 사용해야 함 n번 틀리면 cur + n * (cur + prev

// 또 .. 매개변수.. 탐색.. ?
// 최소는 1로 최대는 diffs의 max값으로 설정

File: 20241022/suhyun/퍼즐게임.java
import java.util.*;

class Solution {
    private static int L;
    
    private long calcTime(int[] diffs, int[] times, int level) {
        long res = times[0];
        
        for (int i = 1; i < L; i++) {
            res += times[i];
            
            if (diffs[i] > level) {
                res += (times[i-1] + times[i]) * (diffs[i] - level);
            }
        }
        
        return res;
    }
    
    public int solution(int[] diffs, int[] times, long limit) {
        L = diffs.length;
        int s = 1;
        int e = 100000;
        int res = Integer.MAX_VALUE;
        
        while (s <= e) {
            int mid = (s + e) / 2;
            long time = calcTime(diffs, times, mid);
            
            if (time > limit) {
                s = mid + 1;
            } else {
                res = Math.min(res, mid);
                e = mid - 1;
            }
        }
        
        return res;
    }
}